03-attribute-operations.Rmd

```diff
@@ -29,12 +39,18 @@ The list elements correspond to the number of observations in the attribute tabl
 But each list element can contain more than one coordinate if required or even another list as it is the case for polygons with holes (see previous sections).
 This structure enables multiple columns to represent a range of attributes for thousands of features (one row per feature).
 
-This chapter focuses on non-geographical operations such as subsetting, aggregating or joining attribute data. 
+This chapter focuses on non-geographical operations on vector and raster data.
+Regarding vector data, we will introduce subsetting, aggregating or joining attribute data. 
 Note that the corresponding functions also have a geographical equivalent.
 Sometimes you can even use the same functions for attribute and spatial operations.
 This is, for example, the case for subsetting as base R's `[` and tidyverse's `filter()` let you also subset spatial data based on the spatial extent of another spatial object (see Chapter \@ref(spatial-data-operations)).
 That is, the skills you learn here are cross-transferable which is also why this chapter lays the foundation for the next chapter (Chapter \@ref(spatial-data-operations)) which extends the here presented methods to the spatial world.
 
+In the case of raster data we will learn how to create, and how to extract cell values from one layer and multiple layers at once (subsetting).
+Finally, we will introduce the concept of map algebra. 
+Since the underlying mathematics is basically matrix algebra (see above), we present them in this chapter (non-spatial attribute operations).
```

Comment:
@Robinlovelace @Nowosad 
I think this is justified but might be controversial. What is your opinion on this matter? Of course, one could also argue to present map algebra in chapter 4. Or we could also think about rearranging the chapter structure. For instance, instead of chapter 3 Non-spatial attributes and 4 Spatial operations, we could have chapter 3 on vector data including non-spatial and spatial operations and chapter 4 on just raster data.

Reply:
That a tough decision to make. Extracting values is definitely "non-spatial" thing, I would assume that zonal operations are "spatial", and map algebra is somewhat in the grey zone... 

@Robinlovelace, what do you think about this?

Reply:
Somewhere above in the text I explain:
`The raster header simply gives the matrix a spatial dimension which we need when plotting the raster or when we want to combine two rasters, think, for instance, of adding the values of one raster to another.
Before such an operation, the corresponding algorithm simply checks if the two rasters share the same extent, projection and resolution, that is the raster headers. 
Hence, the spatial dimension of raster data is (often) of no importance at the processing level.
Then, raster processing uses in fact matrix algebra which is why it is so much faster compared to vector processing.
`
It doesn't matter if is local, focal, zonal or global (all of these operations are part of the map algebra definition), it is just matrix algebra. Zones are not identified via coordinates but via indices, say `mat[c(1:3, 4:7, 12:18), c(24:28, 77:100)]` or cell IDs - nothing spatial about this.
But yes, you are right, might be controversial. On the other hand, when we can explain why we putting it here, you can't really argue with it.
Anyways, it's a tough call as you say, and that's why I also proposed to simply separate vector and raster into two chapters.

Reply:
Interesting discussion and sorry for arriving late to it. I agree that matrix algebra is important when thinking about rasters *in depth* but it's not essential to know that that's why raster datasets are conducive to fast, scalable and parallel processing. I agree that a raster is a very different beast compared with a vector dataset however, having given it some thought, I suggest we don't split-out the raster stuff into a new chapter at this stage.

I think the amount of effort to restructure and re-word everything around this new way of doing things would take-up a fair amount of time, while there are clearly 'lower hanging fruit' in terms of finishing a cohesive draft ready for the reviewers (Chapter 4 is in a sorrier state than Chapter 3 atm, and Chapter 5 could definitely benefit from some work). I think it's a rather 'GIS-y' structure to cover the vector and raster data separately but I think it's good to consider them all as spatial datasets which you can do.

It does make me think about the importance of raster-vector conversion and operations - (see https://stat.ethz.ch/pipermail/r-sig-geo/2017-August/025937.html ) - and suggest we put in a section on Raster-vector operations in Chapter 4 - would fit nicely in there now I've re-ordered things there is plenty of space (I would say 9 sections per chapter is enough).

Reply:
Plus have just added a question to reviewers on this: https://github.com/Robinlovelace/geocompr/commit/de26707cdf1efe927e6da0a8b826842f1bb489b6